### ISSUE TITLE
[Bug][ROCm][P/D] AITER FP8 experts and MLA/MTP decode need MI300X P/D fixes

### ISSUE THREAD

**Issue by chaeminlim-mb** [vllm-project/vllm#45750]
## Summary

MI300X P/D deployments for DeepSeek-style MLA/MoE models need two ROCm AITER paths to be correct:

1. FP8 MoE expert-parallel dispatch must be able to use the `BatchedExperts` activation layout when ROCm AITER is selected.
2. ROCm AITER MLA verification decode must size metadata for the real MTP verification shape and dtype.

This issue links the AITER-related P/D bug fixes and the validation scope. It should not be read as evidence that every MI300X MLA model or every WideEP high-concurrency topology is already passing.

## Your current environment

<details>
<summary>Output of <code>python -m vllm.collect_env</code> (one MI300X node; internal IPs/NICs redacted)</summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.4 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version                : 22.0.0git (https://github.com/RadeonOpenCompute/llvm-project roc-7.2.2 26084 f58b06dce1f9c15707c5f808fd002e18c2accf7e)
CMake version                : version 3.31.10
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.10.0+git8514f05
Is debug build               : False
CUDA used to build PyTorch   : N/A
ROCM used to build PyTorch   : 7.2.53211
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform              : Linux-6.8.0-110-generic-x86_64-with-glibc2.39

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   :
GPU models and configuration : AMD Instinct MI300X (gfx942:sramecc+:xnack-)
Nvidia driver version        : Could not collect
cuDNN version                : Could not collect
HIP runtime version          : 7.2.53211
MIOpen runtime version       : 3.5.1
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  112
On-line CPU(s) list:                     0-111
Vendor ID:                               GenuineIntel
BIOS Vendor ID:                          Intel
Model name:                              INTEL(R) XEON(R) PLATINUM 8570
BIOS Model name:                         INTEL(R) XEON(R) PLATINUM 8570  CPU @ 2.1GHz
BIOS CPU family:                         179
CPU family:                              6
Model:                                   207
Thread(s) per core:                      1
Core(s) per socket:                      56
Socket(s):                               2
Stepping:                                2
CPU(s) scaling MHz:                      46%
CPU max MHz:                             4000.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4200.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities ibpb_exit_to_user
L1d cache:                               5.3 MiB (112 instances)
L1i cache:                               3.5 MiB (112 instances)
L2 cache:                                224 MiB (112 instances)
L3 cache:                                600 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-55
NUMA node1 CPU(s):                       56-111
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

==============================
Versions of relevant libraries
==============================
[pip3] conch-triton-kernels==1.2.1
[pip3] numpy==2.1.3
[pip3] onnx==1.19.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] onnxslim==0.1.94
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+git8514f05
[pip3] torchaudio==2.9.0+eaa9e4e
[pip3] torchvision==0.24.1+d801a34
[pip3] transformers==5.12.1
[pip3] triton==3.7.0
[pip3] triton_kernels==1.0.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : 7.2.53211-35e8c7bf89
vLLM Version                 : 0.1.dev0
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled; XPU: Disabled
GPU Topology:
  ============================ ROCm System Management Interface ============================
================================ Weight between two GPUs =================================
       GPU0         GPU1         GPU2         GPU3         GPU4         GPU5         GPU6         GPU7
GPU0   0            15           15           15           15           15           15           15
GPU1   15           0            15           15           15           15           15           15
GPU2   15           15           0            15           15           15           15           15
GPU3   15           15           15           0            15           15           15           15
GPU4   15           15           15           15           0            15           15           15
GPU5   15           15           15           15           15           0            15           15
GPU6   15           15           15           15           15           15           0            15
GPU7   15           15           15           15           15           15           15           0

================================= Hops between two GPUs ==================================
       GPU0         GPU1         GPU2         GPU3         GPU4         GPU5         GPU6         GPU7
GPU0   0            1            1            1            1            1            1            1
GPU1   1            0            1            1            1            1            1            1
GPU2   1            1            0            1            1            1            1            1
GPU3   1            1            1            0            1            1            1            1
GPU4   1            1            1            1            0            1            1            1
GPU5   1            1            1            1            1            0            1            1
GPU6   1            1            1            1            1            1            0            1
GPU7   1            1            1            1            1            1            1            0

=============================== Link Type between two GPUs ===============================
       GPU0         GPU1         GPU2         GPU3         GPU4         GPU5         GPU6         GPU7
GPU0   0            XGMI         XGMI         XGMI         XGMI         XGMI         XGMI         XGMI
GPU1   XGMI         0            XGMI         XGMI         XGMI         XGMI         XGMI         XGMI
GPU2   XGMI         XGMI         0            XGMI         XGMI         XGMI         XGMI         XGMI
GPU3   XGMI         XGMI         XGMI         0            XGMI         XGMI         XGMI         XGMI
GPU4   XGMI         XGMI         XGMI         XGMI         0            XGMI         XGMI         XGMI
GPU5   XGMI         XGMI         XGMI         XGMI         XGMI         0            XGMI         XGMI
GPU6   XGMI         XGMI         XGMI         XGMI         XGMI         XGMI         0            XGMI
GPU7   XGMI         XGMI         XGMI         XGMI         XGMI         XGMI         XGMI         0

======================================= Numa Nodes =======================================
GPU[0]		: (Topology) Numa Node: 0
GPU[0]		: (Topology) Numa Affinity: 0
GPU[1]		: (Topology) Numa Node: 0
GPU[1]		: (Topology) Numa Affinity: 0
GPU[2]		: (Topology) Numa Node: 0
GPU[2]		: (Topology) Numa Affinity: 0
GPU[3]		: (Topology) Numa Node: 0
GPU[3]		: (Topology) Numa Affinity: 0
GPU[4]		: (Topology) Numa Node: 1
GPU[4]		: (Topology) Numa Affinity: 1
GPU[5]		: (Topology) Numa Node: 1
GPU[5]		: (Topology) Numa Affinity: 1
GPU[6]		: (Topology) Numa Node: 1
GPU[6]		: (Topology) Numa Affinity: 1
GPU[7]		: (Topology) Numa Node: 1
GPU[7]		: (Topology) Numa Affinity: 1
================================== End of ROCm SMI Log ===================================

==============================
     Environment Variables
==============================
VLLM_ROCM_USE_AITER_MLA=True
VLLM_MORIIO_CONNECTOR_READ_MODE=true
TORCHINDUCTOR_CACHE_DIR=/jit-cache/torchinductor
NCCL_IB_DISABLE=0
NCCL_IB_HCA=<redacted>
VLLM_NIXL_SIDE_CHANNEL_HOST=<redacted>
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_WORKER_MULTIPROC_METHOD=spawn
VLLM_ROCM_USE_AITER=1
VLLM_LOGGING_LEVEL=INFO
VLLM_CACHE_ROOT=/jit-cache/vllm
VLLM_MORIIO_NUM_WORKERS=1
VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS=3600
VLLM_HOST_IP=<redacted>
VLLM_NIXL_SIDE_CHANNEL_PORT=5559
VLLM_ROCM_USE_AITER_MOE=True
NCCL_IB_GID_INDEX=3
VLLM_ROCM_USE_AITER_LINEAR=True
VLLM_MORIIO_QP_PER_TRANSFER=4
VLLM_MORIIO_POST_BATCH_SIZE=2
LD_LIBRARY_PATH=/app/deps/install/ucx/lib:/opt/rocm/lib:/usr/local/lib:
PYTORCH_ROCM_ARCH=gfx942
VLLM_TARGET_DEVICE=rocm
CMAKE_BUILD_TYPE=Release
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
```

</details>

vLLM branch under test: `chaemin/upstream-merge-20260611-synced`. The MoRI/MoRIIO/AITER stack is built into the ROCm image from `docker/Dockerfile.rocm`.

### AITER FP8 batched-expert load failure fixed by #45226

Use a DeepSeek-R1 FP8 DP/EP deployment where the FP8 MoE oracle selects `BatchedExperts` activation format and ROCm AITER is enabled (`VLLM_ROCM_USE_AITER=1` and `VLLM_ROCM_USE_AITER_MOE=1`).

Before the fix, the model could fail at load time because the oracle found ROCm AITER, but no ROCm AITER expert implementation advertised support for the batched activation format.

Expected: the model loads and routes the batched expert layout through a correct ROCm AITER path.

### ROCm AITER MLA MTP verification failure fixed by #45227

Use DeepSeek-R1-0528 with ROCm AITER MLA, FP8 KV cache, TP8, and DeepSeek MTP verification (`num_speculative_tokens=1` or `3`). PR #45227 includes builder-level tests and fixes for the metadata sizing path.

Before the fix, metadata could be allocated with stale qlen/dtype assumptions, the persistent qh128 path could fault during warmup, and CUDA-graph padding could create zero-length decode rows.

Expected: MTP verification logits are correct, warmup does not fault on the default path, and CUDA-graph pad rows have valid metadata.

### Reproduction guidance

**Common base flags:**

```bash
--model /path/to/DeepSeek-R1-0528 \
--served-model-name DeepSeek-R1-0528 \
--dtype auto \
--trust-remote-code \
--distributed-executor-backend mp \
--kv-cache-dtype fp8_e4m3 \
--block-size 1 \
--no-enable-prefix-caching \
--compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE"}'
```

**TP8 AITER MLA verification (control):**

```bash
--tensor-parallel-size 8 \
--max-model-len 32768 \
--max-num-batched-tokens 8192 \
--gpu-memory-utilization 0.80 \
--speculative-config '{"method":"deepseek_mtp","num_speculative_tokens":1}'
```

Use `num_speculative_tokens=3` and `gpu-memory-utilization=0.75`, `--max-num-seqs=256` for MTP3.

**DP8EP AITER batched-experts repro:**

```bash
--tensor-parallel-size 1 \
--data-parallel-size 8 \
--enable-expert-parallel \
--all2all-backend mori_high_throughput \
--moe-backend aiter \
--max-model-len 10240 \
--max-num-batched-tokens 8192 \
--gpu-memory-utilization 0.80
```

For wideEP, increase to `--data-parallel-size 16 --data-parallel-size-local 8` across two nodes.

### P/D regression shapes to keep in scope

- TP8 and DP8EP MTP1/MTP3 P/D breadth should remain clean through the tested 1k/1k and 8k/1k rungs.
- WideEP MTP3 `1p1d` 8k/1k c128 is still failing in the broader stack (`101/1280`) and should not be attributed to AITER/MLA unless root cause confirms it.

## Expected behavior

- ROCm AITER can serve FP8 batched expert layouts in DP/EP P/D deployments.
- ROCm AITER MLA MTP verification uses metadata sized for the actual qlen and Q/KV dtype.
- MTP1 and MTP3 validation does not produce corrupted logits, warmup memory faults, or zero-length CUDA-graph pad-row failures.
- AITER/MLA regressions surface directly and are not hidden behind DP finish or shared-memory hangs.

## Actual behavior before the fixes

- AITER FP8 did not have a batched expert implementation for the `BatchedExperts` activation format, so affected DP/EP deployments could refuse to load.
- ROCm AITER MLA MTP verification could use stale qlen/dtype assumptions, causing bad logits or crashes.
- Persistent metadata and CUDA-graph padding paths had additional warmup/pad-row failure modes.

## Related PRs

- https://github.com/vllm-project/vllm/pull/45226 adds `AiterBatchedExpertsFp8`, routes the FP8 oracle to `BATCHED_AITER`, preserves the BatchedExperts router-weight contract, marks converted weights correctly, and adds reshape/oracle tests.
- https://github.com/vllm-project/vllm/pull/45227 fixes ROCm AITER MLA MTP decode metadata sizing for real qlen/dtype, gates the persistent path, fixes CUDA-graph padding metadata, and adds builder-level tests.
- https://github.com/vllm-project/vllm/pull/45223 and https://github.com/vllm-project/vllm/pull/45224 are supporting P/D reliability fixes that make AITER/MLA failures easier to distinguish from DP finish or shared-memory wait issues.

## Evidence and caveats

- Current TP8 and DP8EP P/D coverage is clean through the tested non-relaxed MTP1/MTP3 rungs.
- The tested TP8/DP8EP coverage is consistent with the #45226 AITER batched-experts fix.
- The wider stack still has unresolved WideEP high-concurrency failures. Those failures remain caveats until they are root-caused; this issue should not be used to claim that all WideEP/MI300X MLA deployments pass.
- If the remaining WideEP MTP3 c128 failure is proven to be MoRIIO or shared-memory related rather than AITER/MLA related, this issue can close on the #45226/#45227-specific validation.

## Acceptance criteria

- The #45226 unit tests and FP8 oracle route for `BATCHED_AITER` pass.
- A DeepSeek-R1 FP8 DP/EP deployment that selects `BatchedExperts` can load with ROCm AITER enabled.
- The #45227 ROCm AITER MLA MTP verification tests pass for native and split decode metadata paths.
- TP8 DeepSeek-R1-0528 ROCm AITER MLA MTP1/MTP3 validation produces correct GSM8K-level results without warmup faults.
- P/D TP8 and DP8EP MTP1/MTP3 breadth remains clean after the fixes.
- Remaining WideEP MTP3 failures are either fixed in the AITER/MLA path or explicitly assigned to the MoRIIO/WideEP issue with evidence.

**Comment by chaeminlim-mb** [comment]
Closing this tracking issue for now while we finish internal regression/write-up review. We can reopen or recreate it after the wording and scope are finalized.

**Comment by chaeminlim-mb** [comment]
Reopening with a rewritten, upstream-focused description. The previous body framed this as a generic tracking issue and included internal infra details; this version describes the two concrete AITER/MLA bugs, how the linked PRs fix them, and the remaining WideEP caveats.

**Comment by chaeminlim-mb** [comment]
Tagging co-author @edwinlim0919 (Edwin Lim, MangoBoost), who co-authored the fixes tracked in this issue, so he is looped in and can follow progress here.
